A32-xds-circuit-breaking.md

```diff
@@ -0,0 +1,147 @@
+
+The EDS policy's picker will enforce the configured limit. Each EDS LB policy
+will maintain a counter for the number of requests currently in flight to 
+the associated cluster. The counter will be incremented when a request
+is started on a subchannel and decremented when the request is complete. 
+When the picker is going to start a call, it will check whether the current
+number of in-flight requests is lower than the configured limit. If so, the 
+picker will do the pick as usual, returning a subchannel. If not, it will 
+fail the call with status UNAVAILABLE. Such failed calls will not be retried,
+but they will be recorded to the `total_dropped_requests` counts in 
+cluster-level load reports and reported to the load reporting server.
+
+Note an EDS LB policy may be responsible for handling `service_name` switch
+used for endpoint discovery, requests sent to both the old and new services
+should be aggregated to the same counter and the total is restricted by the
+configured limit. Even though the failed requests will be recorded and reported
+in separate stats message.
+
+## Implementation
+
+### Java
+- Add a `max_requests` field in the EDS LB config. The value will be taken
+from the cluster watch interface pushed by the XdsClient.
+- Maintain a counter in the EDS LB policy for counting the number of 
+outstanding requests.
+- Wrap the `SubchannelPicker` propagated from each of EDS policy's child LB 
+policy with the logic of checking the counter and make pick decision.
+- Counter manipulation is done in `ClientStreamTracer`/`Factory`: increment
+in `newClientStreamTracer()` and decrement in `streamClosed()`.
+    - Note there will be check-and-allocate race as the pick method and stream
+    tracer are not run by the same thread. This will cause the limit to
+    be exceeded. Since the race window is small, the value should be exceeded 
+    only by a small amount.
+
+### C
+TODO
+
+### Go
+TODO
+
+## Other parameters considered
+
+There is a group of other parameters in the xDS circuit breakers, but most of
+them do not apply to gRPC's use case, and they will be ignored in the 
+implementation. The following parameters have been considered:
+
+- `max_connections`: the maximum number of connections can be created to an 
+upstream cluster.
+    - Envoy allows at least one connection to be allocated for each host 
+    selected by cluster load balancing, even if this circuit breaker is
+    overflowed. On the other hand, gRPC creates at most one connection to each
+    host. So this parameter does not have effective meaning in gRPC's use
+    case.
+- `max_pending_requests`: the maximum number of requests that will be queued
+while waiting for a connection to be established.
+    - As a proxy, Envoy drops queued requests from downstream entities to avoid
+    unbounded congestion and memory usage of the process. But as part of the
+    application, it makes little sense for gRPC to drop its own requests.
```

Comment:
Arguably an application that is doing async send would benefit from a cap on otherwise unbounded queueing?

Reply:
@markdroth and I had some discussion about the buffering limit in https://github.com/grpc/proposal/pull/200#pullrequestreview-452627462. I am also leaning towards supporting it, but as mentioned in that comment, it turns out to be more complicated and there are cases we need to consider when putting together with other existing gRPC features.

The reasoning here for not supporting it doesn't seem to be very convincing. Maybe I should say not to be supported in the immediate first round, as it is complicated, and we will be trying to support more complicated ones progressively.

Reply:
Yeah, I think having some distinction between the features that will _never_ be implemented and those that _might_ be implemented later would be great.

Reply:
It wouldn't hurt to expand on the descriptions for these items a bit.  For example, for this particular item, instead of just saying "it makes little sense for gRPC", let's explain that it's complicated to implement and that it's not clear that the benefit is enough to justify the cost at this point.

Reply:
Updated. I also reordered items to put together things not applicable and things with implementation difficulties.